cou/utils/juju_utils.py

```diff
@@ -680,30 +680,26 @@ async def wait_for_idle(  # pylint: disable=too-many-arguments,too-many-position
         """
         if apps is None:
             apps = await self._get_supported_apps()
+        semaphore = asyncio.Semaphore(5)
 
         @retry(timeout=timeout, no_retry_exceptions=(WaitForApplicationsTimeout,))
         @wraps(self.wait_for_idle)
-        async def _wait_for_idle() -> None:
+        async def _wait_for_idle(app) -> None:
             # NOTE(rgildein): Defining wrapper so we can use retry with proper timeout
             model = await self._get_model()
             try:
-                # NOTE(rgildein): Use asyncio.gather because we always go through the application
-                # list (apps will never be None) and libjuju is very slow here.
-                # https://github.com/juju/python-libjuju/issues/1055
-                await asyncio.gather(
-                    *(
-                        model.wait_for_idle(
-                            apps=[app],
-                            timeout=timeout,
-                            idle_period=idle_period,
-                            raise_on_blocked=raise_on_blocked,
-                            raise_on_error=raise_on_error,
-                            status=status,
-                            wait_for_at_least_units=0,  # don't hang if an app has no units
-                        )
-                        for app in apps
+                # NOTE(jneo8): Use semaphore to limit the number of tasks to avoid sending too much
+                # request to API server at the same time.
+                async with semaphore:
+                    await model.wait_for_idle(
+                        apps=[app],
```

Comment:
can't you change single app to all apps?

Reply:
The behaviors are different, details [here](https://github.com/juju/python-libjuju/blob/50b42d013aee01536416e6334f99443f2b4f1e4c/juju/model.py#L2881-L2893)